.github/workflows/test.yml

```diff
@@ -0,0 +1,56 @@
+# This is a GitHub workflow defining a set of jobs with a set of steps.
+# Useful GitHub Actions docs:
+# - https://help.github.com/en/actions
+# - https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions
+# - https://help.github.com/en/actions/configuring-and-managing-workflows/configuring-a-workflow
+# - https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions
+# - https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions
+name: Run tests
+
+# Trigger the workflows on all PRs pushed commits and/or tags
+# to be able to test changes without opening PRs
+on:
+  pull_request:
+  push:
+
+defaults:
+  run:
+    shell: bash
+
+jobs:
+  # Run "pytest tests" for various Python versions
+  pytest:
+    runs-on: ubuntu-20.04
+    strategy:
+      # Keep running even if one variation of the job fail
+      fail-fast: false
+      matrix:
+        tests: [test_installer.py, test_traefik_utils.py, test_traefik_api_auth.py, test_proxy.py]
+        python: ["3.7", "3.8", "3.9"]
+    steps:
+      # NOTE: In GitHub workflows, environment variables are set by writing
+      #       assignment statements to a file. They will be set in the following
+      #       steps as if would used `export MY_ENV=my-value`.
+      - name: Configure environment variables
+        run: |
+          echo "ETCDCTL_API=3" >> $GITHUB_ENV
+          echo "PATH=$PWD/bin:$PATH" >> $GITHUB_ENV
+      - uses: actions/checkout@v2
+      # NOTE: actions/setup-python@v2 make use of a cache within the GitHub base
+      #       environment and setup in a fraction of a second.
+      - name: Install Python ${{ matrix.python }}
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ matrix.python }}
+      - name: Install Python dependencies
+        run: |
+          pip install --upgrade setuptools pip
+          pip install --pre -r dev-requirements.txt --upgrade .
+          python -m jupyterhub_traefik_proxy.install --traefik --etcd --consul --output=./bin
+          pip freeze
+      - name: Run tests
+        run: |
+          pytest -v tests/${{ matrix.tests }} --color=yes --cov=jupyterhub_traefik_proxy
```

Comment:
This makes us run 1 job per test file and python version, why not run all test files here and make it run 1 job per python version?

Reply:
I was experimenting with running them in parallel because I noticed that the proxy and the installer tests take a bit of time. Do you think it looks "hacky" and I shouldn't be doing it?
An example of the jobs that are run:
https://github.com/GeorgianaElena/traefik-proxy/actions/runs/357984398

Reply:
By creating additional jobs, we create additional overhead. So, by splitting it in four, we do the installation of dependencies four times instead of one. In time that adds ~100s or so of total compute time for the entire workflow. We gained in the total time to complete the workflow because we used separate jobs running in parallel without queuing up.

For me, the following things matter:
1. The total time to complete a workflow
2. The ease of understanding how things work quickly and reduce cognitive work for developers
3. The time to run quick test for quick feedback. For example, its good to run the basic linting test before you start the long and slow tests.
4. The total resources used by our CI system
5. The reliability of the outcome assuming things change, where for example github actions reduce our parallel runners to just maximum or similarly.

- If the parallel workers available drop to ~4, then this kind of optimization is wasted.
- The biggest benefit of total time to complete a workflow comes from splitting the test runs into two equally large, so a split of 5:5 rather than 1:1:1:7 for example.

Since the proxy file's tests add up to ~14 minutes and the other tests add up to less than a minute, I suggest to not do this but instead select the tests that are slow with a pytest marker, and then you create two separate pytests jobs. One running `pytest -m slow` and one running `pytest -m "not slow"`. The outcome of this is that you ensure that the fast tests will run and complete in ~1 min, and then for the full test suite to run, you always need to wait a long time.

If you want to optimize under the assumption that tests will keep being able to run in massive parallelism, then you could split apart the tests into X groups of equal size and use that unlimited parallelism to your advantage.

I think it is acceptable as it is though, but I would ask for a comment explaining that the motivation for this is to parallelize some tests and ensure that the quick tests complete quickly and not get stuck behind slow tests.

Reply:
If you mark certain tests as slow, for example all tests in the proxy folder or just some in general, you can sort them with this hook you can defined in conftest.py: https://docs.pytest.org/en/latest/reference.html#pytest.hookspec.pytest_collection_modifyitems, for example based on the markers you have.

Reply:
Thank you for sharing these awesome ideas, Erik. I wasn't entirely sure what I was doing or trying to achieve, other than playing around with gha, jobs and try to make things faster :sweat_smile:

I'm now ordering the tests as per your suggestion, leaving the slow ones at the end, which I think makes a lot of sense to have. Running them in parallel as I did (per test file) wasn't helping much, since the proxy tests are the ones that are slow and numerous, so I was only saving a tini tiny bit of time from running the faster ones in parallel with this.

The best thing to do would probably be splitting the proxy tests somehow, but I'll leave that for another PR.
I believe having them ordered at this point makes more sense. 

TY :heart: